### PR TITLE
Optimizations and build improvements.

### PR DESCRIPTION
**Pull request by adam-azarchs**
NOTE: Absolutely DO NOT merge this before modifying the sake configs to pin the revision being built for PD/CS tarballs for older versions.

A whole bunch of changes:
1.  Code cleanups.
  * Use appropriate C++ data structures, rather than manual memory management.
  * clang-format: no more mixed spaces and tabs.
  * namespaces
  * TSNE shouldn't have been a class - it doesn't have members.
2.  Build improvements.
  * Make a single .so with dynamic dispatch, rather than choosing which dimension mode to use at link time.  Much more reliable and debuggable.
  * Enable LTO
  * Get rid of build dependency on blas/atlas.  It wasn't actually being used, and made it much more complex to package the thing.
   * Break tsne's main, load, or main methods into separate files, which don't need to be built when linking the python extension.
3.  Run-time optimizations.
   * Mostly, reducing the size of the tree nodes by through eliminating redundancy and passing data down from above where appropriate.
   * Flow the templating on dimension further up the stack.
   * Less use needless pointer indirection.
   * Small code restructures to allow better loop vectorization by the compiler.
   * Arithmetic simplifications to eliminate redundant computations.

Testing on cellranger sample 74484, there are some small changes in the output.  I'm not sure if this is because I fixed a bug, introduced a bug, or just a result of numerical instability with some of the changes to arithmetic.

Before:
```
Using random seed: 209652396
Using no_dims = 2, perplexity = 30.000000, and theta = 0.500000
Using max_iter = 1000, stop_lying_iter = 250, mom_switch_iter = 250
Computing input similarities...
Building tree...
 - point 0 of 17030
 - point 10000 of 17030
Input similarities computed in 2.31 seconds (sparsity = 0.007233)!
Learning embedding...
Iteration 50: error is 103.968878 (50 iterations in 3.29 seconds)
Iteration 100: error is 95.993534 (50 iterations in 2.98 seconds)
Iteration 150: error is 84.868840 (50 iterations in 2.63 seconds)
Iteration 200: error is 82.552828 (50 iterations in 2.58 seconds)
Iteration 250: error is 4.310005 (50 iterations in 2.51 seconds)
Iteration 300: error is 3.307660 (50 iterations in 2.29 seconds)
Iteration 350: error is 2.946275 (50 iterations in 2.27 seconds)
Iteration 400: error is 2.718556 (50 iterations in 2.29 seconds)
Iteration 450: error is 2.556061 (50 iterations in 2.30 seconds)
Iteration 500: error is 2.433838 (50 iterations in 2.31 seconds)
Iteration 550: error is 2.337590 (50 iterations in 2.31 seconds)
Iteration 600: error is 2.260225 (50 iterations in 2.32 seconds)
Iteration 650: error is 2.196263 (50 iterations in 2.32 seconds)
Iteration 700: error is 2.143605 (50 iterations in 2.32 seconds)
Iteration 750: error is 2.098874 (50 iterations in 2.32 seconds)
Iteration 800: error is 2.060644 (50 iterations in 2.32 seconds)
Iteration 850: error is 2.027915 (50 iterations in 2.32 seconds)
Iteration 900: error is 1.999645 (50 iterations in 2.32 seconds)
Iteration 950: error is 1.975346 (50 iterations in 2.32 seconds)
Iteration 999: error is 1.955789 (50 iterations in 2.27 seconds)
Fitting performed in 48.59 seconds.
```
After:
```
Using random seed: 209652396
Using D = 10, no_dims = 2, perplexity = 30.000000, and theta = 0.500000
Using max_iter = 1000, stop_lying_iter = 250, mom_switch_iter = 250
Computing input similarities...
Computing approximate perplexity...
Building tree...
 - point 0 of 17030
 - point 10000 of 17030
Input similarities computed in 2.43 seconds (sparsity = 0.007233)!
Learning embedding...
Iteration 50: error is 103.968878 (50 iterations in 2.09 seconds)
Iteration 100: error is 95.993534 (50 iterations in 1.85 seconds)
Iteration 150: error is 84.868779 (50 iterations in 1.61 seconds)
Iteration 200: error is 82.552551 (50 iterations in 1.58 seconds)
Iteration 250: error is 4.310008 (50 iterations in 1.53 seconds)
Iteration 300: error is 3.307299 (50 iterations in 1.39 seconds)
Iteration 350: error is 2.945971 (50 iterations in 1.38 seconds)
Iteration 400: error is 2.715925 (50 iterations in 1.38 seconds)
Iteration 450: error is 2.553057 (50 iterations in 1.39 seconds)
Iteration 500: error is 2.431074 (50 iterations in 1.39 seconds)
Iteration 550: error is 2.335568 (50 iterations in 1.39 seconds)
Iteration 600: error is 2.258516 (50 iterations in 1.39 seconds)
Iteration 650: error is 2.195174 (50 iterations in 1.39 seconds)
Iteration 700: error is 2.142492 (50 iterations in 1.40 seconds)
Iteration 750: error is 2.097975 (50 iterations in 1.41 seconds)
Iteration 800: error is 2.060355 (50 iterations in 1.41 seconds)
Iteration 850: error is 2.027727 (50 iterations in 1.44 seconds)
Iteration 900: error is 1.999039 (50 iterations in 1.42 seconds)
Iteration 950: error is 1.974375 (50 iterations in 1.42 seconds)
Iteration 999: error is 1.953511 (50 iterations in 1.39 seconds)
Fitting performed in 29.65 seconds.
```

### CODE REVIEW

File: tsne/bh_sne_src/sptree.cpp [hunk 1]
```diff
@@ -259,26 +247,32 @@ template<int NDims>
...
     for(unsigned int i = 0; i < size; i++) {
-        bool success = false;
-        for(unsigned int j = 0; j < no_children; j++) {
-            if(!success) success = children[j]->insert(index[i]);
+        for (auto& child : children) {
+            if (child->insert(index[i])) {
+                break;
+            }
         }
-        index[i] = -1;
```
Comment (nlhepler) [inline tsne/bh_sne_src/sptree.cpp]: This line is unreplicated. Is it unnecessary?

File: tsne/bh_sne_src/sptree.cpp [hunk 2]
```diff
@@ -259,26 +247,32 @@ template<int NDims>
...
-            if((i / div) % 2 == 1) new_corner[d] = boundary.getCorner(d) - .5 * boundary.getWidth(d);
-            else                   new_corner[d] = boundary.getCorner(d) + .5 * boundary.getWidth(d);
+            if((i / div) % 2 == 1)
+                new_corner[d] = boundary.getCorner(d) - new_width[d];
+            else
+                new_corner[d] = boundary.getCorner(d) + new_width[d];
             div *= 2;
         }
-        children[i] = new SPTree(this, data, new_corner, new_width);
+        children[i] = unique_ptr<SPTree>(new SPTree(this, data, move(new_corner), child_widths.get()));
```
Comment (nlhepler) [inline tsne/bh_sne_src/sptree.cpp]: Is this safe? You're passing the child_widths to SPTree::init which then sets the raw ptr to a point_t::width pointer--but child_widths will be freed before subdivide() returns.

Comment (adam-azarchs) [reply tsne/bh_sne_src/sptree.cpp]: No, which is one reason why that changed in a later revision.

File: tsne/bh_sne_src/sptree.cpp (hunk 1 above)
Comment (adam-azarchs) [reply tsne/bh_sne_src/sptree.cpp]: Yep.  index is only read for leaf nodes, which this no longer is.

File: tsne/bh_sne_src/tsne.cpp [hunk 3]
```diff
@@ -83,427 +124,427 @@ void TSNE::run(double* X, int N, int D, double* Y, int no_dims, double perplexit
...
+                cur_P[m] = exp(-beta * distances[m + 1] * distances[m + 1]);
 
             // Compute entropy of current row
             sum_P = DBL_MIN;
-            for(int m = 0; m < K; m++) sum_P += cur_P[m];
+            for (int m = 0; m < K; m++) sum_P += cur_P[m];
             double H = .0;
-            for(int m = 0; m < K; m++) H += beta * (distances[m + 1] * distances[m + 1] * cur_P[m]);
+            for (int m = 0; m < K; m++)
+                H += beta * (distances[m + 1] * distances[m + 1] * cur_P[m]);
```
Comment (nlhepler) [inline tsne/bh_sne_src/tsne.cpp]: You could do the same factorization of beta here.

Comment (adam-azarchs) [reply tsne/bh_sne_src/tsne.cpp]: Yeah, there's more that can be done, but 57% of profile samples are on line 264 of sptree.cpp so unless something was offensive to my sense of aesthetics I didn't look too closely.  Will update this, however.

**Comment by adam-azarchs** [comment]
More specifics on the profiling data:
```
55.27% TSNE::_sptree_internal::SPTreeNode<2>::computeNonEdgeForces
16.57% TSNE::run
3.67%  TSNE::_sptree_internal::SPTreeNode<2>::insert
```
The rest is trivial.  Within `computeNonEdgeForces`, the biggest things are:
```asm
       │         if (cum_size == 0 ||
  0.27 │       mov    0x28(%rdi),%eax
 14.48 │       test   %eax,%eax
       │     ↓ je     cc
  1.21 │       cmp    $0x1,%eax
```
and
```asm
       │             double mult = cum_size * sqdist;
  7.65 │       mov    %eax,%eax
```
(which is why the fast-path escape for `cum_size==1`)

The `TSNE::run` samples are actually in `SPTree<2>::computeEdgeForces`, which was inlined, and is as one would expect dominated by the inner loop:
```asm
       │                  const double* data_2 = data + col_P[i] * NDims;
  2.29 │3ab0:   mov    0x148(%rsp),%rsi
  5.82 │        mov    (%r15,%rdi,4),%ebx
  0.65 │        add    %ebx,%ebx
       │                      diffs[d] = data_1[d] - data_2[d];
  0.95 │        movupd (%rax),%xmm1
  2.71 │        movupd (%rsi,%rbx,8),%xmm2
  6.35 │        subpd  %xmm2,%xmm1
       │                      sqdist += diffs[d] * diffs[d];
  0.79 │        movapd %xmm1,%xmm2
  1.00 │        mulpd  %xmm2,%xmm2
  2.94 │        movapd %xmm2,%xmm3
  6.62 │        addsd  %xmm4,%xmm3
  0.51 │        movhlp %xmm2,%xmm2
  1.17 │        addsd  %xmm3,%xmm2
       │                  sqdist = val_P[i] / sqdist;
  2.73 │        movsd  (%r12,%rdi,8),%xmm3
  8.92 │        divsd  %xmm2,%xmm3
       │                      pos_f[d] += sqdist * diffs[d];
  5.33 │        movddu %xmm3,%xmm2
  1.75 │        mulpd  %xmm1,%xmm2
  8.05 │        addpd  %xmm2,%xmm0
 20.55 │        movupd %xmm0,(%rcx)
```
As you can see the for loops don't show up because they've been unrolled and vectorized, which is of course the whole point of templating on `NDims`.

**Comment by adam-azarchs** [comment]
Closing in favor of #10.  Though this does more optimization than that one, that one made its changes more carefully.  Keeping this branch around however for future reference.

**Pull request closed without merging**